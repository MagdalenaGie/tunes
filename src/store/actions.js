import * as actionTypes from './actionTypes';
import axios from '../axios';

const filterAlbumsBy = (filterType, filterValue, albums) => {
    const regex = new RegExp(filterValue, "i");
    let albumsToDisplay = albums.filter( el => el[filterType].match(regex));

    return albumsToDisplay;
}

export const fetchAlbums = (filterType, searchValue) => {
    return dispatch => {
        let route = 'search?term=' + searchValue + '&entity=album';
        axios.get(route)
        .then(res => {
            let albumsToDisplay;
            if(filterType !== "any"){
                albumsToDisplay = filterAlbumsBy(filterType, searchValue, res.data.results);
            }else{
                albumsToDisplay = res.data.results;
            }

            if(albumsToDisplay.length === 0){
                let message = `No albums connected to the term ${searchValue} were found. Please try again with different parameter.`
                dispatch(reportAlbumsFetchError(message));
            }else{
                dispatch(getAlbumsResults(albumsToDisplay));
            }
        })
        .catch(err => {
            console.log(err.message);
            let message = `A probblem has occured while connecting to the database, please try again later`;
            dispatch(reportAlbumsFetchError(message));
        })
    }
}

export const reportAlbumsFetchError = (err) => {
    return{
        type: actionTypes.ALBUMS_ERR,
        err: err
    }
}

export const getAlbumsResults = (results) => {
    return{
        type: actionTypes.SEARCH_FOR_ALBUMS,
        results: results
    }
}

export const fetchSongsFromAlbum = (albumId) => {
    return dispatch => {
        let route = 'lookup?id=' + albumId + '&entity=song';
        axios.get(route)
        .then(res => {
            dispatch(getSongsFromAlbum(res.data.results));
        })
        .catch(err => {
            console.log(err.message);
            let message = `A probblem has occured while connecting to the database, please try again later`;
            dispatch(reportSongsFetchError(message))
        })
    }
}

export const reportSongsFetchError = (err) => {
    return{
        type: actionTypes.SONG_ERR,
        err: err
    }
}

export const getSongsFromAlbum = (songs) => {
    songs.shift();
    return{
        type: actionTypes.GET_SONGS_FROM_ALBUM,
        songs: songs
    }
}

export const resetSongs = () => {
    return{
        type: actionTypes.RESET_SONGS
    }
}