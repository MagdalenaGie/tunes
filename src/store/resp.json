{
    "resultCount":1,
    "results": [
   {"wrapperType":"collection", "collectionType":"Album", "artistId":201246132, "collectionId":668699145, "amgArtistId":19398, "artistName":"The Monks", "collectionName":"Black Monk Time", "collectionCensoredName":"Black Monk Time", "artistViewUrl":"https://music.apple.com/us/artist/the-monks/201246132?uo=4", "collectionViewUrl":"https://music.apple.com/us/album/black-monk-time/668699145?uo=4", "artworkUrl60":"https://is3-ssl.mzstatic.com/image/thumb/Music6/v4/9a/39/5e/9a395e3e-ef06-23c8-3e26-6aca7bf7562d/source/60x60bb.jpg", "artworkUrl100":"https://is3-ssl.mzstatic.com/image/thumb/Music6/v4/9a/39/5e/9a395e3e-ef06-23c8-3e26-6aca7bf7562d/source/100x100bb.jpg", "collectionPrice":9.99, "collectionExplicitness":"notExplicit", "trackCount":18, "copyright":"â„— 2009 Light In The Attic Records", "country":"USA", "currency":"USD", "releaseDate":"2009-04-14T07:00:00Z", "primaryGenreName":"Rock"}]
}