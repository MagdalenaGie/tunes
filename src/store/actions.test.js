import moxios from 'moxios';
import axios from './../axios';
import * as actions from './actions';
import resp from './resp.json';
import songResp from './songresp.json';

test('getAlbumsResults', () => {
    expect(actions.getAlbumsResults(resp.results)).toMatchSnapshot();
})

test('getSongsFromAlbum', () => {
    expect(actions.getSongsFromAlbum([{type: 'album - to be shifted'}, {type: 'song - will be saved'}])).toMatchSnapshot();
})

test('resetSongs', () => {
    expect(actions.resetSongs()).toMatchSnapshot();
})

test('reportSongsFetchError', () => {
    expect(actions.reportSongsFetchError('error message')).toMatchSnapshot();
})

test('reportAlbumsFetchError', () => {
    expect(actions.reportAlbumsFetchError('error message')).toMatchSnapshot();
})

test('fetchAlbums', done => {
    moxios.install(axios);
    const dispatchMock = jest.fn();
    moxios.withMock(() => {
        actions.fetchAlbums("collectionName", "black")(dispatchMock);
        moxios.wait(() => {
            let request = moxios.requests.mostRecent();
            request.respondWith({
                status: 200,
                response: resp
            })
            .then(() => {
                expect(request.url).toEqual('search?term=black&entity=album');
                expect(dispatchMock).toBeCalledWith(actions.getAlbumsResults(resp.results));
                done();
            })
        })
    })
})

test('fetchSongsFromAlbum', done => {
    moxios.install(axios);
    const dispatchMock = jest.fn();
    moxios.withMock(() => {
        actions.fetchSongsFromAlbum("1545341237")(dispatchMock);
        moxios.wait(() => {
            let request = moxios.requests.mostRecent();
            request.respondWith({
                status: 200,
                response: songResp
            })
            .then(() => {
                expect(request.url).toEqual('lookup?id=1545341237&entity=song');
                expect(dispatchMock).toBeCalledWith(actions.getSongsFromAlbum(songResp.results));
                done();
            })
        })
    })
})


