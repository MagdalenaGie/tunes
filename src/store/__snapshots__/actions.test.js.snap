// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAlbumsResults 1`] = `
Object {
  "results": Array [
    Object {
      "amgArtistId": 19398,
      "artistId": 201246132,
      "artistName": "The Monks",
      "artistViewUrl": "https://music.apple.com/us/artist/the-monks/201246132?uo=4",
      "artworkUrl100": "https://is3-ssl.mzstatic.com/image/thumb/Music6/v4/9a/39/5e/9a395e3e-ef06-23c8-3e26-6aca7bf7562d/source/100x100bb.jpg",
      "artworkUrl60": "https://is3-ssl.mzstatic.com/image/thumb/Music6/v4/9a/39/5e/9a395e3e-ef06-23c8-3e26-6aca7bf7562d/source/60x60bb.jpg",
      "collectionCensoredName": "Black Monk Time",
      "collectionExplicitness": "notExplicit",
      "collectionId": 668699145,
      "collectionName": "Black Monk Time",
      "collectionPrice": 9.99,
      "collectionType": "Album",
      "collectionViewUrl": "https://music.apple.com/us/album/black-monk-time/668699145?uo=4",
      "copyright": "â„— 2009 Light In The Attic Records",
      "country": "USA",
      "currency": "USD",
      "primaryGenreName": "Rock",
      "releaseDate": "2009-04-14T07:00:00Z",
      "trackCount": 18,
      "wrapperType": "collection",
    },
  ],
  "type": "SEARCH_FOR_ALBUMS",
}
`;

exports[`getSongsFromAlbum 1`] = `
Object {
  "songs": Array [
    Object {
      "type": "song - will be saved",
    },
  ],
  "type": "GET_SONGS_FROM_ALBUM",
}
`;

exports[`reportAlbumsFetchError 1`] = `
Object {
  "err": "error message",
  "type": "ALBUMS_ERR",
}
`;

exports[`reportSongsFetchError 1`] = `
Object {
  "err": "error message",
  "type": "SONG_ERR",
}
`;

exports[`resetSongs 1`] = `
Object {
  "type": "RESET_SONGS",
}
`;
